File: server/internal/domain/container.go
package domain

import (
	"errors"
	"time"

	"github.com/google/uuid"
)

// Container errors.
var (
	ErrContainerNotFound                            = errors.New("container not found")                     // Returned when a container is not found.
	ErrContainerAssociatedWithContainerReport       = errors.New("container associated with user report")   // Returned when a container is associated with a user report.
	ErrContainerAssociatedWithUserContainerBookmark = errors.New("container associated with user bookmark") // Returned when a container is associated with a user bookmark.
	ErrContainerAssociatedWithRouteContainer        = errors.New("container associated with route")         // Returned when a container is associated with a route.
)

// ContainerCategory defines the category of the container.
type ContainerCategory string

const (
	ContainerCategoryGeneral   ContainerCategory = "general"
	ContainerCategoryPaper     ContainerCategory = "paper"
	ContainerCategoryPlastic   ContainerCategory = "plastic"
	ContainerCategoryMetal     ContainerCategory = "metal"
	ContainerCategoryGlass     ContainerCategory = "glass"
	ContainerCategoryOrganic   ContainerCategory = "organic"
	ContainerCategoryHazardous ContainerCategory = "hazardous"
)

// Valid returns true if the category is valid, false otherwise.
func (c ContainerCategory) Valid() bool {
	switch c {
	case ContainerCategoryGeneral,
		ContainerCategoryPaper,
		ContainerCategoryPlastic,
		ContainerCategoryMetal,
		ContainerCategoryGlass,
		ContainerCategoryOrganic,
		ContainerCategoryHazardous:
		return true
	default:
		return false
	}
}

// EditableContainer defines the editable container structure.
type EditableContainer struct {
	Category ContainerCategory
	GeoJSON  GeoJSON
}

// EditableContainerPatch defines the patchable container structure.
type EditableContainerPatch struct {
	Category *ContainerCategory
	GeoJSON  GeoJSON
}

// Container defines the container structure.
type Container struct {
	EditableContainer
	ID         uuid.UUID
	CreatedAt  time.Time
	ModifiedAt time.Time
}

// ContainerPaginatedSort defines the field of the container to sort.
type ContainerPaginatedSort string

const (
	ContainerPaginatedSortCategory         ContainerPaginatedSort = "category"
	ContainerPaginatedSortWayName          ContainerPaginatedSort = "wayName"
	ContainerPaginatedSortMunicipalityName ContainerPaginatedSort = "municipalityName"
	ContainerPaginatedSortCreatedAt        ContainerPaginatedSort = "createdAt"
	ContainerPaginatedSortModifiedAt       ContainerPaginatedSort = "modifiedAt"
)

// Field returns the name of the field to sort by.
func (s ContainerPaginatedSort) Field() ContainerPaginatedSort {
	return s
}

// Valid returns true if the field is valid, false otherwise.
func (s ContainerPaginatedSort) Valid() bool {
	switch s {
	case ContainerPaginatedSortCategory,
		ContainerPaginatedSortWayName,
		ContainerPaginatedSortMunicipalityName,
		ContainerPaginatedSortCreatedAt,
		ContainerPaginatedSortModifiedAt:
		return true
	default:
		return false
	}
}

// ContainersPaginatedFilter defines the containers filter structure.
type ContainersPaginatedFilter struct {
	PaginatedRequest[ContainerPaginatedSort]
	Category         *ContainerCategory
	WayName          *string
	MunicipalityName *string
}

File: android/app/src/main/java/com/ecomap/ecomap/domain/Container.kt
package com.ecomap.ecomap.domain

import android.content.Context
import com.ecomap.ecomap.R

/**
 * Represents all the available container categories.
 */
enum class ContainerCategory {
    GENERAL,
    PAPER,
    PLASTIC,
    METAL,
    GLASS,
    ORGANIC,
    HAZARDOUS;

    /**
     * Returns the respective drawable icon resource.
     * @return Resource ID.
     */
    fun getIconResource(): Int {
        return when (this) {
            GENERAL -> R.drawable.general
            PAPER -> R.drawable.paper
            PLASTIC -> R.drawable.plastic
            METAL -> R.drawable.metal
            GLASS -> R.drawable.glass
            ORGANIC -> R.drawable.organic
            HAZARDOUS -> R.drawable.hazardous
        }
    }

    /**
     * Returns the respective human readable string resource.
     * @param context Activity context.
     * @return String resource.
     */
    fun getStringResource(context: Context): String {
        return when (this) {
            GENERAL -> context.getString(R.string.container_category_general)
            PAPER -> context.getString(R.string.container_category_paper)
            PLASTIC -> context.getString(R.string.container_category_plastic)
            METAL -> context.getString(R.string.container_category_metal)
            GLASS -> context.getString(R.string.container_category_glass)
            ORGANIC -> context.getString(R.string.container_category_organic)
            HAZARDOUS -> context.getString(R.string.container_category_hazardous)
        }
    }
}

/**
 * Represents the structure of a container.
 * @param id Container id.
 * @param category Container category.
 * @param geoJSON Container location in the GeoJSON format.
 * @param createdAt Timestamp of when the container was created.
 * @param modifiedAt Timestamp of when the container was modified.
 */
data class Container(
    val id: String,
    val category: ContainerCategory,
    val geoJSON: GeoJSONFeaturePoint,
    val createdAt: String,
    val modifiedAt: String,
)

/**
 * Represents the structure of paginated containers.
 * @param total The total amount of resources available for the provided filter.
 * @param containers The list of containers for the provided filter.
 */
data class ContainersPaginated(
    val total: Int,
    val containers: ArrayList<Container>
)
